// ezdev-pcview/core/render/analyzer/RouterAnalyzer.js
/**
 * Created by zhaofs on 2016/7/21.
 * --------------------
 * - 路由分析器 -
 * --------------------
 * 本模块主要实现类似<a>标签中的href属性功能，只要在此元素点时会把e-router属性指定的href值
 * 设置到浏览器的url中，以便触发路由事件。
 * e-router属性的格式为"href:/test/crud/search?name=zhao&url=/crud/users/pages"
 */
import config from "../../Config";
import htmlUtils from "../../utils/HtmlUtils";
import eventUtils from "../../utils/EvenUtils";
import BaseAnalyzer from "./BaseAnalyzer";

export default class extends BaseAnalyzer {
    constructor() {
        super();
        this.name = "RouterAnalyzer";
    }

    /**
     * 对元素进行分析
     * @param content
     */
    analyze(content) {
        let handles = $(content).find('[' + config.routerTag + ']');
        handles.each(function (index, item) {
            $(item).on('click', function (event) {
                let resolveRouter = this.resolveRouter($(event.currentTarget));
                eventUtils.setCurrentEventTarget($(event.currentTarget));
                window.location.href = '/#' + resolveRouter.href;
            }.bind(this));

        }.bind(this));
    }

    /**
     * 分析路由相关信息
     * @param item
     */
    resolveRouter(item) {
        let router = htmlUtils.getJsonByStr(item.attr(config.routerTag));
        return {
            href: router.href
        }
    }

};
