// ezdev-pcview/core/utils/MiscUtils.js
/**
 * Created by zhaofs on 2017/5/27.
 * --------------------
 * - 汇合工具 -
 * --------------------
 * 各种常用工具大结合。
 */

//add lhw 20160809
export default {
    instanceHash : {},
    /**
     * 获取唯一实例
     * @param className
     */
    getInstance: function (className) {
        if (this.instanceHash[className]) {
            return this.instanceHash[className];
        } else {
            return undefined;
        }
    },
    /**
     * 把实例加入hash表
     * @param className
     * @param instance
     */
    setInstance: function (className, instance) {
        this.instanceHash[className] = instance;
    },
    //add end


    //ADD LHW 20151126 追加获取uuid函数，用于生成唯一的id。 START
    guid: function () {
        return 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx'.replace(/[x]/g, function (c) {
            let r = Math.random() * 16 | 0, v = c == 'x' ? r : (r & 0x3 | 0x8);
            return v.toString(16);
        });
    },
    //ADD LHW 20151126 END
    initdate: function (args) {
        if (!args || !args.length) {
            return;
        }
        for (let i = 0; i < args.length; i++) {
            let id = args[i];
            $("#" + id).datepicker({format: 'yyyy-mm-dd', autoclose: true, language: 'zh-CN', todayHighlight: true});
            $("#" + id).attr("readonly", "readonly");
            $("#" + id).css("background-color", "#ffffff");
            $("#" + id).css("cursor", "pointer");
        }
    },
    //add lhw 20170217 模块片段http接口调用功能（工作流中的表单使用）
    analyzeWindowHref: function () {

        let windowHref = {};
        //原始href：localhost/#FullScreen/tpl/demo/message/url/fcc/sqb/c?id=xxxxx
        //转换为：FullScreen/tpl/demo/message/url/fcc/sqb/c?id=xxxxx
        windowHref.href = this.getAnchorUrl();

        //如果href为全屏展示模块片段，向util中放入所有href信息。
        if (windowHref.href && windowHref.href.indexOf("FullScreen") === 0) {
            windowHref.isFullScreen = true;

            let tplAndUrl = windowHref.href.substr(11);//形如：tpl/demo/message/url/fcc/sqb/c?id=xxxxx
            let urlIndex = tplAndUrl.indexOf("/url/");
            if (urlIndex > 0) {
                windowHref.fsHasUrl = true;
                windowHref.fsUrl = tplAndUrl.substr(urlIndex + 5);
                windowHref.fsTpl = tplAndUrl.substring(4, urlIndex);
            } else {
                windowHref.fsHasUrl = false;
                windowHref.fsTpl = tplAndUrl.substr(4);
            }
        } else {
            windowHref.isFullScreen = false;
        }

        this.setInstance("windowHref", windowHref);
    },
    getWindowHref: function () {
        return this.getInstance("windowHref");
    },
    //add end
};
