// ezdev-pcview/webpack.config.js
var path = require("path");
module.exports = {
    entry: './example/index.js',
    output: {
        filename: 'bundle.js',
        path: path.resolve(__dirname, '../dist/static/dist'),
        publicPath: '/dist/'
    },
    module: {

        rules: [{
            test: /\.html$/,
            loader: "art-template-loader",
            options: {
                // art-template options (if necessary)
                // imports: require.resolve("./template-runtime"),
                htmlResourceRules: false
            }
        }, {
            test: /\.js$/,
            exclude: /(node_modules)/,
            use: [{
                loader: 'babel-loader',
                options: {
                    presets: [['es2015', {modules: false}]],
                    plugins: [
                        'syntax-dynamic-import'
                    ]
                }
            }]
        },{
            test: /test\.js$/,
            use: 'mocha-loader',
            exclude: /node_modules/
        }]
    }

};
