// ezdev-pcview/core/components/ComponentManager.js
/**
 * Created by wangchen on 2017/7/31.
 * --------------------
 * - html组件管理 -
 * --------------------
 * 实现对系统调用组件的统一管理
 * renderHtml方法为常用方法，只需要传入符合的“e-”开头的dom元素即可，目前在组件扫描分析器中配合使用
 */
import ESelect from "./select/ESelect";
import ERadio from "./radio/ERadio";
import ECheckbox from "./checkbox/ECheckbox";
import EDaterangepicker from "./date/EDaterangepicker";
import EDatetimepicker from "./date/EDatetimepicker";
import EPopu from "./popu/EPopu";
import popuHandler from "./popu/Handler";
import handlerManager from "../handler/HandlerManager";


const comsMap = new Map();
export default {
    /**
     * 初始化框架中的基本组件
     */
    initComponents: function() {
        this.add(ESelect.instance());
        this.add(ERadio.instance());
        this.add(ECheckbox.instance());
        this.add(EDaterangepicker.instance());
        this.add(EDatetimepicker.instance());
        this.add(EPopu.instance());
        // let registName = "core/components/html/popu";
        //let registName = "/core/components/html/popu";
        handlerManager.staticImport({"/component/popu":popuHandler});
    },
    /**
     * 添加组件
     */
    add: function(com) {
        if (typeof(com) != 'undefined' && com != null && com != '') {
            comsMap.set(com.name, com);
            /*let registName = "core/components/html/"+com.eventName;
            handlerManager.staticImport({registName:com});*/
        } else {
            throw new Error("你添加的组件不符合要求，无法进行添加");
        }
    },

    /**
     * 获取所有分析器
     * @returns {Array}
     */
    getAll: function() {
        return [...comsMap.values()];
    },

    getAllComNames: function() {
        return [...comsMap.keys()];
    },

    /**
     * 是否存在
     * @param name
     */
    has: function(name) {
        return comsMap.has(name);
    },
    /**
     * 根据name获取组件对象
     */
    get: function(name) {
        return comsMap.get(name);
    },

    /**
     * 根据dom元素获取组件对象
     */
    getByDom: function(dom) {
        let ccName = this.getTagClassName(dom);
        return this.get(ccName);
    },

    /**
     * 根据标签获取对应的组件class名称
     */
    getTagClassName(dom) {
        let tg = $(dom).prop('tagName');
        // let className = tg.toLowerCase().replace('e-','');
        let className = tg.toLowerCase();
        return className;
    },
    //首字母大写
    capUpperCase(str) {
        return str.toLowerCase().replace(/( |^)[a-z]/g, (L) => L.toUpperCase());
    },
    /**
     * 将传入的dom元素根据标签自动调用相应的组件进行html渲染
     * @para:dom
     */
    renderHtml(dom) {
        let com = this.getByDom(dom);
        let args = {
            el: dom
        };
        com.reWriteHtml(args);
    }
};
